compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -214,7 +214,9 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
                     struct_span_err!(tcx.sess, attr.span, E0737, "`#[track_caller]` requires Rust ABI")
                         .emit();
                 }
-                if tcx.is_closure(did.to_def_id()) && !tcx.features().closure_track_caller {
+                if !tcx.features().async_fn_track_caller
+                    && (tcx.is_closure(did.to_def_id()) && !tcx.features().closure_track_caller)
```

Comment:
You should use `mark_span_with_reason`/`allow_internal_unstable` (which is carried in a `Span`) so that lowering usages of `#[feature(async_fn_track_caller)]` into usages of `#[feature(closure_track_caller)]` doesn't result in additional feature errors. 

This check for `tcx.features().async_fn_track_caller` will need to be or'd with a call to `span.allows_unstable(sym::closure_track_caller)` or whatever, I think.